### ISSUE TITLE
Process should be able to give its subprocess a tty for its standard output

### ISSUE THREAD

**Issue by kulakowski** [dart-lang/sdk#24716]
Without this facility, certain kinds of tools are problematic to write in Dart. Anything that wants to be a better shell script, essentially, falls down, as all the invoked processes behave poorly and format their output differently.

**Comment by sethladd** [comment]
cc @nex3

**Comment by nex3** [comment]
Signal-handling in nested processes is also impossible to get right in general.

One possible general solution for these issues would be to provide a wrapper for the POSIX `exec()` function (and whatever the Windows equivalent is). This *replaces* the current process with a different one, which ensures that the nested process behaves just like the parent (since it's not actually nested).

**Comment by sgjesse** [comment]
It the issue here that the subprocess see a pipe device for its stdout/stderr instead of a terminal device - and therefore might behave differently? Do you have a concrete example of a program where this can be reproduced?

So instead of always using pipes to child processed we should also have an option for allocating a pty and use its master/slave fds for connecting the processes?

**Comment by kulakowski** [comment]
@sgjesse Yes that's precisely the issue.

A good example for us is invoking [ninja](https://github.com/martine/ninja/blob/a48434493d69085aa6fc0c002180ace555cad592/src/line_printer.cc#L33), in which link you can see the |isatty| check to configure line pretty printing. On a tty, it overwrites progress lines. Not on a tty, it prints everything on a fresh line.

We previously had a mix of Makefile wrappers and python subprocess.call()s for this which were correctly handling this.

**Comment by sgjesse** [comment]
I experimented a bit with this, and providing a pty to a spawned process is not that complicated. However, the pty master file descriptor is not supported by either `epoll` on Linux nor `kqueue` on Mac OS. This means that the current eventhandler cannot be used for waiting on events on the pty master fd. As far as I can a separate eventhandler thread using `select` will have to be added somehow to give access to the pty.

**Comment by natebosch** [comment]
Should we consider this a dup of https://github.com/dart-lang/sdk/issues/23165 or vice-versa?

**Comment by zanderso** [comment]
As stated, this issue is resolved by https://github.com/dart-lang/sdk/commit/6523896c6fb27a7ead2ced038da4d15164d894d4. Some of the discussion in this issue overlaps with https://github.com/dart-lang/sdk/issues/23165, but the change doesn't address ptys, so I'll close this one, and leave the other open. We probably need a separate issue for the signals/execve use-case.

**Comment by kevmoo** [comment]
Thanks, @zanderso !!

**Comment by kevmoo** [comment]
The change to add ProcessStartMode.INHERIT_STDIO is now in 2.0.0-dev.20.0 https://github.com/dart-lang/sdk/commit/6523896c6fb27a7ead2ced038da4d15164d894d4
